src/librustdoc/clean/mod.rs

```diff
@@ -579,8 +583,61 @@ fn is_elided_lifetime(param: &hir::GenericParam<'_>) -> bool {
 
 pub(crate) fn clean_generics<'tcx>(
     gens: &hir::Generics<'tcx>,
+    def_id: DefId,
     cx: &mut DocContext<'tcx>,
+    is_in_impl: bool,
 ) -> Generics {
+    let mut bounds_to_remove: FxHashMap<Type, (Vec<GenericBound>, Vec<Lifetime>)> =
+        Default::default();
+    let mut regions_to_remove: FxHashMap<Lifetime, Vec<GenericBound>> = Default::default();
+
+    // There can't be duplicates between the item's generics and its parent's if we're not in an
+    // impl block.
+    if is_in_impl {
+        // FIXME: Once we can differentiate between generics (`T: Foo`) and where predicates
+        // (`where T: Foo`), we should remove the `gens` argument and directly build from the
+        // results returned here.
+        let substs = InternalSubsts::identity_for_item(cx.tcx, def_id);
+        let predicates =
+            cx.tcx.explicit_predicates_of(def_id).instantiate(cx.tcx, substs).predicates;
+        let param_env = cx.tcx.param_env(cx.tcx.parent(def_id));
+        let param_env = EarlyBinder(param_env).subst(cx.tcx, substs);
+
+        for predicate in predicates {
```

Comment:
I think you could also register all obligations in one go and then just look at the ones in the error list to see which ones not to remove

Reply:
I told Guillaume to not do this because the `resolve_regions` is destructive to the inference context, and it's very likely to give you an error back that's not possible to correlate back to a bound.

Reply:
I actually picked the opposite logic because we still need to merge bounds split between multiple parts like this:

```rust
fn foo<T: Bar>()
     where T: Foo, T: Foo2
```

It seemed simpler to instead check which ones to remove since we still need to go through the `hir::Generics`. Once we got rid of the `hir::Generics` argument of the function, all this logic will be merged directly into this loop since we won't need multiple passes for it. Or did I completely misunderstood what you meant?

Reply:
I added the explanations from @compiler-errors.